packages/flutter_tools/lib/src/ios/code_signing.dart

```diff
@@ -116,229 +159,791 @@ Future<Map<String, String>?> getCodeSigningIdentityDevelopmentTeamBuildSetting({
+              ?.group(1);
+        })
+        .where(_isNotEmpty)
+        .whereType<String>()
+        .toSet() // Unique.
+        .toList();
+  }
 
-  // If `security`'s output format changes, we'd have to update the above regex.
-  if (signingCertificateId == null) {
+  /// Validates the saved provisioning profile still exists and that there is a
+  /// valid identity/certificate for the profile.
+  ///
+  /// Returns null if profile cannot be found, parsed, or validated.
+  Future<_ProvisioningProfile?> _validateSavedProfile(
+    String savedProfilePath,
+    List<String> validCodeSigningIdentities,
+  ) async {
+    final File savedProfile = _fileSystem.file(savedProfilePath);
+    if (!savedProfile.existsSync()) {
+      _logger.printError('Unable to find saved provisioning profile $savedProfilePath');
+      return null;
+    }
+    final _ProvisioningProfile? parsedProfile = await _parseProvisioningProfile(savedProfile);
+    if (parsedProfile == null) {
+      return null;
+    }
+    for (final File cert in parsedProfile.developerCertificates) {
+      final String? identity = await _validateIdentityFromCert(cert, validCodeSigningIdentities);
+      if (identity != null) {
+        return parsedProfile;
+      }
+    }
+    _logger.printError(
+      'Unable to find a valid certificate matching the provisioning profile $savedProfilePath',
+    );
     return null;
   }
 
-  String signingCertificateStdout;
-  try {
-    signingCertificateStdout =
-        (await processUtils.run(<String>[
-          'security',
-          'find-certificate',
-          '-c',
-          signingCertificateId,
-          '-p',
-        ], throwOnError: true)).stdout.trim();
-  } on ProcessException catch (error) {
-    logger.printTrace("Couldn't find the certificate: $error.");
-    return null;
+  /// Decode and convert a .mobileprovision file to a .plist file and then
+  /// parse the .plist into [_ProvisioningProfile].
+  Future<_ProvisioningProfile?> _parseProvisioningProfile(File provisioningProfileFile) async {
+    final Directory profilesDirectory = _fileSystem.systemTempDirectory.childDirectory(
+      'provisioning_profiles',
+    );
+    profilesDirectory.createSync(recursive: true);
+    final File decodedProfile = profilesDirectory.childFile(
+      'decoded_profile_${provisioningProfileFile.basename}.plist',
+    );
+    try {
+      await _processUtils.run(<String>[
+        'security',
+        'cms',
+        '-D',
+        '-i',
+        provisioningProfileFile.path,
+        '-o',
+        decodedProfile.path,
+      ], throwOnError: true);
+    } on ProcessException catch (error) {
+      _logger.printError('Unexpected failure from security: $error.');
+      return null;
+    }
+    if (!decodedProfile.existsSync()) {
+      _logger.printError('Failed to decode ${provisioningProfileFile.basename}');
+      return null;
+    }
+    final Map<String, Object> contents = _plistParser.parseFile(decodedProfile.path);
```

Comment:
Should this be in the try/catch in case the decoded profile contains invalid XML?

Reply:
`plutil`, which `_plistParser` uses will fail if it's invalid: 

https://github.com/flutter/flutter/blob/bf3826d2ce928dc8ed0d66c0aeb4c8c3a86705ab/packages/flutter_tools/lib/src/ios/plist_parser.dart#L48-L67

Reply:
Right, but `PlistParser._parseXml` then attempts to decode the XML outputted by `plutil`. If for whatever reason `plutil` had unexpected output, that would throw. FWIW, I don't feel strongly about this, it seems like an edge case